src/wp-admin/includes/image.php

```diff
@@ -251,9 +253,11 @@ function wp_create_image_subsizes( $file, $attachment_id ) {
 	if ( $exif_meta ) {
 		$image_meta['image_meta'] = $exif_meta;
 	}
-
+	$mime_type = $imagesize['mime'];
 	// Do not scale (large) PNG images. May result in sub-sizes that have greater file size than the original. See #48736.
-	if ( 'image/png' !== $imagesize['mime'] ) {
+	if ( 'image/png' !== $mime_type ) {
+		$valid_mime_transforms = wp_upload_image_mime_transforms( $attachment_id );
+		$output_mime_types     = isset( $valid_mime_transforms[ $mime_type ] ) ? $valid_mime_transforms[ $mime_type ] : array( $mime_type );
```

Comment:
See my above comment, I think it would be valuable to split this array into the primary MIME type and the additional MIME types to generate. For example something like this:
```suggestion
		$output_mime_types     = isset( $valid_mime_transforms[ $mime_type ] ) ? $valid_mime_transforms[ $mime_type ] : array( $mime_type );

		// Use original mime type as primary mime type, or alternatively the first one.
		$primary_mime_type_key = array_search( $mime_type, $output_mime_types, true );
		if ( false === $primary_mime_type_key ) {
			$primary_mime_type_key = 0;
		}
		// Split output mime types into primary mime type and additional mime types.
		$additional_mime_types     = $output_mime_types;
		list( $primary_mime_type ) = array_splice( $additional_mime_types, $primary_mime_type_key, 1 );
```

Then anywhere below, we can first run the applicable logic for the primary MIME type and then (if there are any) for the additional MIME types. The logic is always slightly different, so having the MIME types separate avoids the needs for if clauses.

Reply:
Ah, that works! I guess I was wondering if there is a use case for having the primary images be in a different mime type than the upload, while still generating more than one type (eg upload jpeg, primary type is webp, secondary type is jpeg. we can leave off supporting that though, this makes the logic about which is primary easier to understand.

Reply:
Applied this, thanks!